File: src/test/java/com/joecollins/bindings/BindingTest.java
package com.joecollins.bindings;

import static org.junit.Assert.assertEquals;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import org.apache.commons.lang3.mutable.Mutable;
import org.apache.commons.lang3.mutable.MutableObject;
import org.junit.Test;

public class BindingTest {

  @Test
  public void testFixedBinding() {
    Mutable<Integer> boundValue = new MutableObject<>();
    Binding<Integer> binding = Binding.fixedBinding(42);
    binding.bind(boundValue::setValue);
    assertEquals(42, boundValue.getValue().intValue());
  }

  @Test
  public void testPropertyBinding() {
    Mutable<Integer> boundValue = new MutableObject<>();
    BindableInt bindable = new BindableInt(7);
    Binding<Integer> binding =
        Binding.propertyBinding(bindable, BindableInt::getValue, BindableInt.Property.VALUE);
    binding.bind(boundValue::setValue);
    assertEquals(7, boundValue.getValue().intValue());

    bindable.setValue(42);
    assertEquals(42, boundValue.getValue().intValue());

    binding.unbind();
    bindable.setValue(7);
    assertEquals(42, boundValue.getValue().intValue());
  }

  @Test
  public void testSizeBinding() {
    Mutable<Integer> boundValue = new MutableObject<>();
    BindableList<Integer> list = new BindableList<>();
    list.add(7);
    Binding<Integer> binding = Binding.sizeBinding(list);
    binding.bind(boundValue::setValue);
    assertEquals(1, boundValue.getValue().intValue());

    list.add(42);
    assertEquals(2, boundValue.getValue().intValue());

    list.remove(1);
    assertEquals(1, boundValue.getValue().intValue());

    list.add(0, 7);
    assertEquals(2, boundValue.getValue().intValue());

    list.clear();
    assertEquals(0, boundValue.getValue().intValue());

    list.add(7);
    assertEquals(1, boundValue.getValue().intValue());

    binding.unbind();
    list.add(42);
    assertEquals(1, boundValue.getValue().intValue());
  }

  @Test
  public void testMapBinding() {
    Mutable<String> boundValue = new MutableObject<>();
    BindableInt bindable = new BindableInt(7);
    Binding<String> binding =
        Binding.propertyBinding(bindable, BindableInt::getValue, BindableInt.Property.VALUE)
            .map(Object::toString);
    binding.bind(boundValue::setValue);
    assertEquals("7", boundValue.getValue());

    bindable.setValue(42);
    assertEquals("42", boundValue.getValue());

    binding.unbind();
    bindable.setValue(7);
    assertEquals("42", boundValue.getValue());
  }

  @Test
  public void testMergeBinding() {
    Mutable<Integer> boundValue = new MutableObject<>();
    BindableInt bindable1 = new BindableInt(7);
    BindableInt bindable2 = new BindableInt(42);
    Binding<Integer> binding1 =
        Binding.propertyBinding(bindable1, BindableInt::getValue, BindableInt.Property.VALUE);
    Binding<Integer> binding2 =
        Binding.propertyBinding(bindable2, BindableInt::getValue, BindableInt.Property.VALUE);
    Binding<Integer> binding = binding1.merge(binding2, Integer::sum);
    binding.bind(boundValue::setValue);
    assertEquals(49, boundValue.getValue().intValue());

    bindable1.setValue(42);
    assertEquals(84, boundValue.getValue().intValue());

    bindable2.setValue(7);
    assertEquals(49, boundValue.getValue().intValue());

    binding.unbind();
    bindable1.setValue(7);
    assertEquals(49, boundValue.getValue().intValue());
    bindable2.setValue(1);
    assertEquals(49, boundValue.getValue().intValue());
  }

  @Test
  public void testBindingReceiverBasic() {
    Mutable<Integer> boundValue1 = new MutableObject<>();
    Mutable<Integer> boundValue2 = new MutableObject<>();
    BindableInt bindable = new BindableInt(7);
    Binding<Integer> binding =
        Binding.propertyBinding(bindable, BindableInt::getValue, BindableInt.Property.VALUE);
    BindingReceiver<Integer> receiver = new BindingReceiver<>(binding);
    Binding<Integer> binding1 = receiver.getBinding();
    binding1.bind(boundValue1::setValue);
    Binding<Integer> binding2 = receiver.getBinding();
    binding2.bind(boundValue2::setValue);
    assertEquals(7, boundValue1.getValue().intValue());
    assertEquals(7, boundValue2.getValue().intValue());

    bindable.setValue(42);
    assertEquals(42, boundValue1.getValue().intValue());
    assertEquals(42, boundValue2.getValue().intValue());

    binding1.unbind();
    bindable.setValue(7);
    assertEquals(42, boundValue1.getValue().intValue());
    assertEquals(7, boundValue2.getValue().intValue());
  }

  @Test
  public void testBindingReceiverMap() {
    Mutable<Integer> boundValue1 = new MutableObject<>();
    Mutable<Integer> boundValue2 = new MutableObject<>();
    BindableInt bindable = new BindableInt(7);
    Binding<Integer> binding =
        Binding.propertyBinding(bindable, BindableInt::getValue, BindableInt.Property.VALUE);
    BindingReceiver<Integer> receiver = new BindingReceiver<>(binding);
    Binding<Integer> binding1 = receiver.getBinding(i -> i * 2);
    binding1.bind(boundValue1::setValue);
    Binding<Integer> binding2 = receiver.getBinding(i -> i * i);
    binding2.bind(boundValue2::setValue);
    assertEquals(14, boundValue1.getValue().intValue());
    assertEquals(49, boundValue2.getValue().intValue());

    bindable.setValue(42);
    assertEquals(84, boundValue1.getValue().intValue());
    assertEquals(1764, boundValue2.getValue().intValue());

    binding1.unbind();
    bindable.setValue(7);
    assertEquals(84, boundValue1.getValue().intValue());
    assertEquals(49, boundValue2.getValue().intValue());
  }

  @Test
  public void testBindingReceiverFlatMap() {
    Mutable<Integer> boundValue = new MutableObject<>();
    BindableInt bindable1 = new BindableInt(7);
    NestedBindable<Integer> nested = new NestedBindable<>(bindable1);
    Binding<BindableValue<Integer>> nestedBinding =
        Binding.propertyBinding(nested, NestedBindable::getValue, BindableInt.Property.VALUE);
    BindingReceiver<BindableValue<Integer>> receiver = new BindingReceiver<>(nestedBinding);
    Binding<Integer> binding =
        receiver.getFlatBinding(
            v -> Binding.propertyBinding(v, BindableValue::getValue, BindableInt.Property.VALUE));
    binding.bind(boundValue::setValue);
    assertEquals(7, boundValue.getValue().intValue());

    bindable1.setValue(42);
    assertEquals(42, boundValue.getValue().intValue());

    BindableInt bindable2 = new BindableInt(12);
    nested.setValue(bindable2);
    assertEquals(12, boundValue.getValue().intValue());

    bindable1.setValue(1);
    assertEquals(12, boundValue.getValue().intValue());

    binding.unbind();
    bindable2.setValue(10);
    assertEquals(12, boundValue.getValue().intValue());

    nested.setValue(new BindableInt(27));
    assertEquals(12, boundValue.getValue().intValue());
  }

  @Test
  public void testMapReduceBinding() {
    Mutable<Integer> boundValue = new MutableObject<>();
    List<BindableInt> list =
        Arrays.asList(new BindableInt(1), new BindableInt(2), new BindableInt(3));
    List<Binding<Integer>> bindings =
        list.stream()
            .map(
                b ->
                    Binding.propertyBinding(
                        b, BindableValue::getValue, BindableValue.Property.VALUE))
            .collect(Collectors.toList());
    Binding<Integer> binding =
        Binding.mapReduceBinding(bindings, 0, (a, v) -> a + v, (a, v) -> a - v);
    binding.bind(boundValue::setValue);
    assertEquals(6, boundValue.getValue().intValue());

    list.get(0).setValue(4);
    list.get(1).setValue(5);
    list.get(2).setValue(6);
    assertEquals(15, boundValue.getValue().intValue());

    binding.unbind();

    list.get(0).setValue(1);
    list.get(1).setValue(2);
    list.get(2).setValue(3);
    assertEquals(15, boundValue.getValue().intValue());
  }

  private static class BindableValue<T> extends Bindable {
    enum Property {
      VALUE
    }

    private T value;

    public BindableValue() {}

    public BindableValue(T value) {
      setValue(value);
    }

    public T getValue() {
      return value;
    }

    public void setValue(T value) {
      this.value = value;
      onPropertyRefreshed(Property.VALUE);
    }
  }

  private static class BindableInt extends BindableValue<Integer> {
    public BindableInt() {
      super();
    }

    public BindableInt(int value) {
      super(value);
    }
  }

  private static class NestedBindable<T> extends BindableValue<BindableValue<T>> {
    public NestedBindable() {
      super();
    }

    public NestedBindable(BindableValue<T> value) {
      super(value);
    }
  }
}

File: src/test/java/com/joecollins/bindings/BindingTest.kt
package com.joecollins.bindings

import com.joecollins.bindings.Binding.Companion.fixedBinding
import com.joecollins.bindings.Binding.Companion.listBinding
import com.joecollins.bindings.Binding.Companion.mapReduceBinding
import com.joecollins.bindings.Binding.Companion.propertyBinding
import com.joecollins.bindings.Binding.Companion.sizeBinding
import com.joecollins.graphics.utils.BoundResult
import org.junit.Assert
import org.junit.Test

class BindingTest {

    @Test
    fun testFixedBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val binding = fixedBinding(42)
        binding.bind { boundValue.value = it }
        Assert.assertEquals(42, boundValue.value.toLong())
    }

    @Test
    fun testPropertyBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val bindable = BindableInt(7)
        val binding = propertyBinding(bindable, { it.value }, BindableValue.Property.VALUE)
        binding.bind { boundValue.value = it }
        Assert.assertEquals(7, boundValue.value.toLong())
        bindable.setValue(42)
        Assert.assertEquals(42, boundValue.value.toLong())
        binding.unbind()
        bindable.setValue(7)
        Assert.assertEquals(42, boundValue.value.toLong())
    }

    @Test
    fun testSizeBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val list = BindableList<Int>()
        list.add(7)
        val binding = sizeBinding(list)
        binding.bind { boundValue.value = it }
        Assert.assertEquals(1, boundValue.value.toLong())
        list.add(42)
        Assert.assertEquals(2, boundValue.value.toLong())
        list.removeAt(1)
        Assert.assertEquals(1, boundValue.value.toLong())
        list.add(0, 7)
        Assert.assertEquals(2, boundValue.value.toLong())
        list.clear()
        Assert.assertEquals(0, boundValue.value.toLong())
        list.add(7)
        Assert.assertEquals(1, boundValue.value.toLong())
        binding.unbind()
        list.add(42)
        Assert.assertEquals(1, boundValue.value.toLong())
    }

    @Test
    fun testMapBinding() {
        val boundValue: BoundResult<String> = BoundResult()
        val bindable = BindableInt(7)
        val binding = propertyBinding(bindable, { it.value }, BindableValue.Property.VALUE)
                .map { it.toString() }
        binding.bind { boundValue.value = it }
        Assert.assertEquals("7", boundValue.value)
        bindable.setValue(42)
        Assert.assertEquals("42", boundValue.value)
        binding.unbind()
        bindable.setValue(7)
        Assert.assertEquals("42", boundValue.value)
    }

    @Test
    fun testMergeBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val bindable1 = BindableInt(7)
        val bindable2 = BindableInt(42)
        val binding1 = propertyBinding(bindable1, { it.value }, BindableValue.Property.VALUE)
        val binding2 = propertyBinding(bindable2, { it.value }, BindableValue.Property.VALUE)
        val binding = binding1.merge(binding2) { a, b -> a + b }
        binding.bind { boundValue.value = it }
        Assert.assertEquals(49, boundValue.value.toLong())
        bindable1.setValue(42)
        Assert.assertEquals(84, boundValue.value.toLong())
        bindable2.setValue(7)
        Assert.assertEquals(49, boundValue.value.toLong())
        binding.unbind()
        bindable1.setValue(7)
        Assert.assertEquals(49, boundValue.value.toLong())
        bindable2.setValue(1)
        Assert.assertEquals(49, boundValue.value.toLong())
    }

    @Test
    fun testBindingReceiverBasic() {
        val boundValue1: BoundResult<Int> = BoundResult()
        val boundValue2: BoundResult<Int> = BoundResult()
        val bindable = BindableInt(7)
        val binding = propertyBinding(bindable, { it.value }, BindableValue.Property.VALUE)
        val receiver = BindingReceiver(binding)
        val binding1 = receiver.getBinding()
        binding1.bind { boundValue1.value = it }
        val binding2 = receiver.getBinding()
        binding2.bind { boundValue2.value = it }
        Assert.assertEquals(7, boundValue1.value.toLong())
        Assert.assertEquals(7, boundValue2.value.toLong())
        bindable.setValue(42)
        Assert.assertEquals(42, boundValue1.value.toLong())
        Assert.assertEquals(42, boundValue2.value.toLong())
        binding1.unbind()
        bindable.setValue(7)
        Assert.assertEquals(42, boundValue1.value.toLong())
        Assert.assertEquals(7, boundValue2.value.toLong())
    }

    @Test
    fun testBindingReceiverMap() {
        val boundValue1: BoundResult<Int> = BoundResult()
        val boundValue2: BoundResult<Int> = BoundResult()
        val bindable = BindableInt(7)
        val binding = propertyBinding(bindable, { it.value }, BindableValue.Property.VALUE)
        val receiver = BindingReceiver(binding)
        val binding1 = receiver.getBinding { it * 2 }
        binding1.bind { boundValue1.value = it }
        val binding2 = receiver.getBinding { it * it }
        binding2.bind { boundValue2.value = it }
        Assert.assertEquals(14, boundValue1.value.toLong())
        Assert.assertEquals(49, boundValue2.value.toLong())
        bindable.setValue(42)
        Assert.assertEquals(84, boundValue1.value.toLong())
        Assert.assertEquals(1764, boundValue2.value.toLong())
        binding1.unbind()
        bindable.setValue(7)
        Assert.assertEquals(84, boundValue1.value.toLong())
        Assert.assertEquals(49, boundValue2.value.toLong())
    }

    @Test
    fun testBindingReceiverFlatMap() {
        val boundValue: BoundResult<Int> = BoundResult()
        val bindable1 = BindableInt(7)
        val nested = NestedBindable(bindable1)
        val nestedBinding = propertyBinding(nested, { it.value }, BindableValue.Property.VALUE)
        val receiver = BindingReceiver(nestedBinding)
        val binding = receiver.getFlatBinding { propertyBinding(it, { e -> e.value }, BindableValue.Property.VALUE) }
        binding.bind { boundValue.value = it }
        Assert.assertEquals(7, boundValue.value.toLong())
        bindable1.setValue(42)
        Assert.assertEquals(42, boundValue.value.toLong())
        val bindable2 = BindableInt(12)
        nested.setValue(bindable2)
        Assert.assertEquals(12, boundValue.value.toLong())
        bindable1.setValue(1)
        Assert.assertEquals(12, boundValue.value.toLong())
        binding.unbind()
        bindable2.setValue(10)
        Assert.assertEquals(12, boundValue.value.toLong())
        nested.setValue(BindableInt(27))
        Assert.assertEquals(12, boundValue.value.toLong())
    }

    @Test
    fun testMapReduceBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val list = listOf(BindableInt(1), BindableInt(2), BindableInt(3))
        val bindings = list
                .map { b: BindableInt ->
                    propertyBinding(b, { it.value }, BindableValue.Property.VALUE)
                }
                .toList()
        val binding = mapReduceBinding(bindings, 0, { a: Int, v: Int -> a + v }, { a: Int, v: Int -> a - v })
        binding.bind { boundValue.value = it }
        Assert.assertEquals(6, boundValue.value.toLong())
        list[0].setValue(4)
        list[1].setValue(5)
        list[2].setValue(6)
        Assert.assertEquals(15, boundValue.value.toLong())
        binding.unbind()
        list[0].setValue(1)
        list[1].setValue(2)
        list[2].setValue(3)
        Assert.assertEquals(15, boundValue.value.toLong())
    }

    @Test
    fun testListBinding() {
        val boundValue: BoundResult<List<Int>> = BoundResult()
        val list = listOf(BindableInt(1), BindableInt(2), BindableInt(3))
        val bindings = list
            .map { b: BindableInt ->
                propertyBinding(b, { it.value }, BindableValue.Property.VALUE)
            }
            .toList()
        val binding = listBinding(bindings)
        binding.bind { boundValue.value = it }
        Assert.assertEquals(listOf(1, 2, 3), boundValue.value)
        list[0].setValue(4)
        list[1].setValue(5)
        list[2].setValue(6)
        Assert.assertEquals(listOf(4, 5, 6), boundValue.value)
        binding.unbind()
        list[0].setValue(1)
        list[1].setValue(2)
        list[2].setValue(3)
        Assert.assertEquals(listOf(4, 5, 6), boundValue.value)
    }

    @Test
    fun testEmptyMapReduceBinding() {
        val boundValue: BoundResult<Int> = BoundResult()
        val bindings: List<Binding<Int>> = listOf()
        val binding = mapReduceBinding(bindings, 1, { a, b -> a + b }, { a, b -> a - b })
        binding.bind { boundValue.value = it }
        Assert.assertEquals(1, boundValue.value.toLong())
    }

    private open class BindableValue<T>(value: T) : Bindable<BindableValue<T>, BindableValue.Property>() {
        enum class Property {
            VALUE
        }

        var value: T = value
            private set

        fun setValue(value: T) {
            this.value = value
            onPropertyRefreshed(Property.VALUE)
        }
    }

    private class BindableInt(value: Int) : BindableValue<Int>(value)

    private class NestedBindable<T>(value: BindableValue<T>) : BindableValue<BindableValue<T>>(value)
}
